File: week_2/day_08_stack/next_greater_element_stack.py
# File: next_greater_element_stack.py

def nextGreater(nums):
    stack = []
    result = [-1] * len(nums)

    for i in range(len(nums)):
        while stack and nums[i] > nums[stack[-1]]:
            idx = stack.pop()
            result[idx] = nums[i]
        stack.append(i)

    return result


# INPUT
nums = [4, 5, 2, 25]
print(nextGreater(nums))  # Output: [5, 25, 25, -1]

File: week_2/day_08_stack/nextGreaterElementStack.js
// File: nextGreaterElementStack.js

function nextGreater(nums) {
    const stack = [];
    const res = new Array(nums.length).fill(-1);

    for (let i = 0; i < nums.length; i++) {
        while (stack.length && nums[i] > nums[stack[stack.length - 1]]) {
            const idx = stack.pop();
            res[idx] = nums[i];
        }
        stack.push(i);
    }
    return res;
}

// INPUT
console.log(nextGreater([4, 5, 2, 25])); // [5, 25, 25, -1]
